{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Import packages\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import multiprocessing\n",
    "import tqdm\n",
    "import time\n",
    "from Utils import ss_utils as ss \n",
    "from Utils import ss_database as db"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Download COVID-19 US data with NCBI Accession - And calculate codon biases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Previous_Data = False \n",
    "\n",
    "CPU_USED = 20\n",
    "\n",
    "# Load data downloaded from NCBI\n",
    "df_download = pd.read_csv('1. Downloaded raw data//NCBI_Virus_Screenshots//sequences.csv')\n",
    "Index_Accession_List = list(enumerate(df_download['Accession']))\n",
    "\n",
    "Run_List = list(range(len(Index_Accession_List)))\n",
    "\n",
    "# Main program for downloading Genome Length\n",
    "def PROCESS(RunList_Index):\n",
    "\n",
    "    index, accession_id = Index_Accession_List[RunList_Index]\n",
    "\n",
    "    all_cds_detail_dict = ss.get_all_cds_detail(accession_id)   # Download CDS details\n",
    "\n",
    "    cds_len_list = []\n",
    "    conten_cds = ''\n",
    "    for gene_name, cds_detail_dict in all_cds_detail_dict.items():\n",
    "\n",
    "        cds_seq = cds_detail_dict['mRNA_sequence']\n",
    "\n",
    "        if len(cds_seq) % 3 == 0:\n",
    "            cds_len_list.append(len(cds_seq))\n",
    "            conten_cds += cds_seq   # Make contenCDS\n",
    "        else:\n",
    "            cds_len_list = []\n",
    "            conten_cds = ''\n",
    "            break\n",
    "\n",
    "    # Codon Usage\n",
    "    cu = ss.codon_usage(conten_cds)   # Calculate Codon Usages\n",
    "\n",
    "    # Codon Usage Sum\n",
    "    cu_sum = sum(list(cu.values()))  \n",
    "\n",
    "    # RSCU\n",
    "    rscu = ss.codon_usage_to_RSCU_usage(cu)   # Calculate RSCU\n",
    "\n",
    "    # CDS Len\n",
    "    dis_data = ss.discriptive_data(cds_len_list)\n",
    "    cds_len_dict = {'contenCDS_length'   :dis_data['sum'],\n",
    "                    'cds_count'          :dis_data['count'],\n",
    "                    'cds_length_mean'    :dis_data['mean'],\n",
    "                    'cds_length_std'     :dis_data['std'],\n",
    "                    'cds_length_median'  :dis_data['median'],\n",
    "                    'cds_length_variance':dis_data['variance']}\n",
    "    \n",
    "    return RunList_Index, cu, rscu, cds_len_dict, accession_id, cu_sum\n",
    "\n",
    "# Create output folder\n",
    "foldername = '1. Downloaded raw data//COVID19'\n",
    "ss.create_folder(foldername)\n",
    "\n",
    "if Previous_Data is False:\n",
    "    Final_CU_Dict, Final_RSCU_Dict, Final_CDS_Len_Dict, Accession_ID_Dict, Final_CU_Sum_Dict = {}, {}, {}, {}, {}\n",
    "\n",
    "# Multiprocessing program\n",
    "POOL = multiprocessing.Pool(processes = CPU_USED)\n",
    "\n",
    "for _RESULT_ in tqdm.tqdm(POOL.imap_unordered(PROCESS, Run_List), \n",
    "                          total = len(Run_List), miniters=50):\n",
    "    \n",
    "    RunList_Index, covid_cu_dict, covid_rscu_dict, covid_cds_len_dict, accession_id, cu_sum = _RESULT_\n",
    "    \n",
    "    Final_CU_Dict[RunList_Index] = covid_cu_dict\n",
    "    Final_RSCU_Dict[RunList_Index] = covid_rscu_dict\n",
    "    Final_CDS_Len_Dict[RunList_Index] = covid_cds_len_dict\n",
    "    Accession_ID_Dict[RunList_Index] = accession_id\n",
    "    Final_CU_Sum_Dict[RunList_Index] = cu_sum\n",
    "\n",
    "    # Temporally Save data \n",
    "    if RunList_Index % 200 == 0:\n",
    "        \n",
    "        ss.database_save(Final_CU_Dict, f\"{foldername}//Final_CU_Dict\", 'Output')\n",
    "        ss.database_save(Final_RSCU_Dict, f\"{foldername}//Final_RSCU_Dict\", 'Output')\n",
    "        ss.database_save(Final_CDS_Len_Dict, f\"{foldername}//Final_CDS_Len_Dict\", 'Output')\n",
    "        ss.database_save(Accession_ID_Dict, f\"{foldername}//Accession_ID_Dict\", 'Output')\n",
    "        ss.database_save(Final_CU_Sum_Dict, f\"{foldername}//CU_Sum_Dict\", 'Output')\n",
    "\n",
    "# Convert to dataframe and save data \n",
    "ss.database_save(Final_CU_Dict, f\"{foldername}//Final_CU_Dict\", 'Output')\n",
    "ss.database_save(Final_RSCU_Dict, f\"{foldername}//Final_RSCU_Dict\", 'Output')\n",
    "ss.database_save(Final_CDS_Len_Dict, f\"{foldername}//Final_CDS_Len_Dict\", 'Output')\n",
    "ss.database_save(Accession_ID_Dict, f\"{foldername}//Accession_ID_Dict\", 'Output')\n",
    "ss.database_save(Final_CU_Sum_Dict, f\"{foldername}//CU_Sum_Dict\", 'Output')\n",
    "DF_Final_CU = pd.DataFrame.from_dict(Final_CU_Dict, 'index')\n",
    "DF_Final_RSCU = pd.DataFrame.from_dict(Final_RSCU_Dict, 'index')\n",
    "DF_Final_CDS_Len = pd.DataFrame.from_dict(Final_CDS_Len_Dict, 'index')\n",
    "DF_Final_Accession_ID = pd.DataFrame.from_dict(Accession_ID_Dict, 'index')\n",
    "DF_Final_CU_Sum = pd.DataFrame.from_dict(Final_CU_Sum_Dict, 'index')\n",
    "ss.save_file(DF_Final_CU, f\"{foldername}//DF_Final_CU\")\n",
    "ss.save_file(DF_Final_RSCU, f\"{foldername}//DF_Final_RSCU\")\n",
    "ss.save_file(DF_Final_CDS_Len, f\"{foldername}//DF_Final_CDS_Len\")\n",
    "ss.save_file(DF_Final_Accession_ID, f\"{foldername}//DF_Final_Accession_ID\")\n",
    "ss.save_file(DF_Final_CU_Sum, f\"{foldername}//DF_Final_CU_Sum\")\n",
    "\n",
    "print ('Done and save.')\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Download Genome Length (required for CDS_Length dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Input previous downloaded data\n",
    "Previous_Data = False\n",
    "Final_Genome_Len = {}\n",
    "\n",
    "\n",
    "COVID_Accession_ID = pd.read_csv(f\"1. Downloaded raw data//COVID19//Final_Data//DF_Final_Accession_ID.csv\")\n",
    "Index_Accession_List = list(enumerate(COVID_Accession_ID['id']))\n",
    "Run_List = list(range(len(Index_Accession_List)))\n",
    "\n",
    "if Previous_Data is True:\n",
    "    Run_List = Run_List[Index_Out:]\n",
    "    print (f\"Run_List rearranged at Index-{Index_Out}.\")\n",
    "else:\n",
    "    pass\n",
    "\n",
    "foldername = f\"1. Downloaded raw data//COVID19\"\n",
    "CPU_USED = 20\n",
    "\n",
    "\n",
    "# Main program for downloading Genome Length\n",
    "def PROCESS(RunList_Index):\n",
    "    index, accession_id = Index_Accession_List[RunList_Index]\n",
    "    Download_Trial = 0\n",
    "    while  Download_Trial < 3:\n",
    "        try:\n",
    "            seq = ss.get_whole_seq(accession_id)[accession_id]   # Download whole sequences\n",
    "            break\n",
    "        except:\n",
    "            time.sleep(1)\n",
    "            Download_Trial += 1\n",
    "            seq = ''\n",
    "            \n",
    "    genome_length = len(seq)\n",
    "    output = {'genome_length':genome_length}\n",
    "    return RunList_Index, output\n",
    "\n",
    "# Multiprocessing program\n",
    "POOL = multiprocessing.Pool(processes = CPU_USED)\n",
    "for _RESULT_ in tqdm.tqdm(POOL.imap_unordered(PROCESS, Run_List), \n",
    "                          total = len(Run_List), miniters=50):\n",
    "    \n",
    "    RunList_Index, output = _RESULT_\n",
    "    Final_Genome_Len[RunList_Index] = output\n",
    "    \n",
    "    # Temporally Save data \n",
    "    if RunList_Index % 200 == 0:\n",
    "        ss.database_save(Final_Genome_Len, f\"{foldername}//Final_Genome_Len\", 'Output')\n",
    "        \n",
    "# Convert to dataframe and save data \n",
    "ss.database_save(Final_Genome_Len, f\"{foldername}//Final_Genome_Len\", 'Output')\n",
    "DF_Final_Genome_Len = pd.DataFrame.from_dict(Final_Genome_Len, 'index')\n",
    "ss.save_file(DF_Final_Genome_Len, f\"{foldername}//DF_Final_Genome_Len\")\n",
    "print ('Done and save.')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Organise into 'RTG_Data'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Input downloaded final data\n",
    "RSCU = ss.load_RTG_Data(\"RSCU\")\n",
    "TAXONOMY = ss.load_RTG_Data(\"TAXONOMY\")\n",
    "CDS_LENGTH = ss.load_RTG_Data(\"CDS_LENGTH\")\n",
    "INFO = ss.load_RTG_Data(\"INFO\")\n",
    "\n",
    "COVID_RSCU = pd.read_csv(f\"1. Downloaded raw data//COVID19//Raw_Data//DF_Final_RSCU.csv\")\n",
    "COVID_CU_Sum = pd.read_csv(f\"1. Downloaded raw data//COVID19//Raw_Data//DF_Final_CU_Sum.csv\")\n",
    "COVID_CU = pd.read_csv(f\"1. Downloaded raw data//COVID19//Raw_Data//DF_Final_CU.csv\")\n",
    "COVID_CDS_Len = pd.read_csv(f\"1. Downloaded raw data//COVID19//Raw_Data//DF_Final_CDS_Len.csv\")\n",
    "COVID_Accession_ID = pd.read_csv(f\"1. Downloaded raw data//COVID19//Raw_Data//DF_Final_Accession_ID.csv\")\n",
    "COVID_Genome_Len = pd.read_csv(f\"1. Downloaded raw data//COVID19/Raw_Data//DF_Final_Genome_Len.csv\")\n",
    "\n",
    "\n",
    "# Organise RSCU data to X-ready\n",
    "COVID_RSCU = COVID_RSCU[db.multibox_codon_table_list]\n",
    "ref_rscu_col = {}\n",
    "for codon in db.multibox_codon_table_list:\n",
    "    ref_rscu_col[codon] = f\"[RSCU]_{codon}\"\n",
    "COVID_RSCU = COVID_RSCU.rename(columns=ref_rscu_col)\n",
    "print (f\"RSCU columns check = {COVID_RSCU.columns.tolist() == RSCU.columns.tolist()}\")\n",
    "\n",
    "# Organise Taxonomy data to X-ready\n",
    "COVID_Name = 'Severe acute respiratory syndrome coronavirus 2'\n",
    "COVID_Index = (INFO['id'] == COVID_Name)\n",
    "COVID_TAXONOMY = TAXONOMY[COVID_Index].reset_index(drop=True)\n",
    "COVID_TAXONOMY = COVID_TAXONOMY.loc[COVID_TAXONOMY.index.repeat(COVID_Accession_ID.shape[0])].reset_index(drop=True)\n",
    "print (f\"TAXONOMY columns check = {COVID_TAXONOMY.columns.tolist() == TAXONOMY.columns.tolist()}\")\n",
    "\n",
    "# Organise CDS_Len data to X-ready\n",
    "CDS_LENGTH = ss.load_RTG_Data(\"CDS_LENGTH\")\n",
    "CDS_Len_col_list = CDS_LENGTH.columns.tolist()\n",
    "COVID_CDS_Len = pd.concat([COVID_Genome_Len, COVID_CDS_Len[CDS_Len_col_list[1:]]], axis=1)\n",
    "print (f\"CDS_Len columns check = {COVID_CDS_Len.columns.tolist() == CDS_LENGTH.columns.tolist()}\")\n",
    "\n",
    "\n",
    "# Save into dataframes\n",
    "foldername = '1. Downloaded raw data//COVID19//RTG_Data'\n",
    "ss.create_folder(foldername)\n",
    "ss.save_file(COVID_RSCU, f\"{foldername}//RSCU\")\n",
    "ss.save_file(COVID_TAXONOMY, f\"{foldername}//TAXONOMY\")\n",
    "ss.save_file(COVID_CDS_Len, f\"{foldername}//CDS_LENGTH\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
